Fix rotation of selection below 0

chrisduerr opened the pull request:
Whenever the viewport is scrolled, the selection is rotated to make sure
that it moves with the viewport. However this did not correctly handle
the underflow that happens when the selection goes below 0.

This resolves that problem for the selection by moving the internal line
representation to an isize, thus correctly keeping track of the
selection start/end points even when they have a negative index. Once
the selection is converted to a span, the lines are clamped to the
visible region.

This fixes #1640 and fixes #1643.

Note that this currently breaks semantic and line selection because
the necessary adjustments to handle `isize`s haven't been made in their
span methods yet. This should be fairly simple, however opening a
temporary PR here should prevent unnecessary duplicated effort.

chrisduerr commented:
I was looking into this again today and it turns out that the semantic selection is currently very broken in Alacritty in the alternate screen buffer. This includes the problem that when scrolling the selection off the screen (to the top), that the top line is only partially selected. Another issue is that copying to the clipboard does not go all the way.

Since the causes for this are a little more complicated, this will probably need a bit more work. Currently the semantic selection is updated quite frequently whenever the selection is converted to a span. I would assume that this requires to be changed.

zacps commented on src/index.rs:
```diff
@@ -52,6 +52,18 @@ impl Ord for Point {
     }
 }
 
+impl From<Point<usize>> for Point<isize> {
```
Is there any reason we use both `usize` and `isize` instead of just `isize`?

chrisduerr replied:
Well generally the index should never be negative. And it actually never is negative in the normal screen buffer. So we use `usize` for most stuff internally.

But with the alternate screen buffer, scrolling doesn't actually scroll a buffer, there's not actually anything going up/down, because the alternate screen buffer has only the visible region. As a result the line `0` is always the bottom, so if we want to keep track of everything below that, using negative lines seems reasonable. Thus the `isize` stuff.

It doesn't make much sense to do this for the complete codebase though, since most scenarios never want to keep track of anything outside of the terminal emulator. So for all other places the only thing relevant is the offset from the bottom of all lines, which is `0`. So a `usize` works perfectly.

I've decided to use `event.rs` as a hand-off point for this since it is an intermediate step between `input.rs` and `selection.rs`. So everything talking to it can just use the standard `usize` which are used everywhere and then the selection can take over which actually has a concept of lines below the terminal emulator.

Note that this PR still needs another clean-up pass. So some things might seem nonsensical (haven't looked at the resulting diff yet). But the general idea shouldn't be totally off.

chrisduerr commented:
Cleaned things up a bit and added some tests to prevent potential regressions.

If all tests pass, this should be good to go as far as I'm concerned.
